xds/src/main/java/io/grpc/xds/EnvoyProtoData.java

```diff
@@ -135,17 +175,589 @@ public String toString() {
+   * Boolean}, or {@code null}.
+   */
+  private static Value convertToValue(Object rawObject) {
+    Value.Builder valueBuilder = Value.newBuilder();
+    if (rawObject == null) {
+      valueBuilder.setNullValue(NullValue.NULL_VALUE);
+    } else if (rawObject instanceof Double) {
+      valueBuilder.setNumberValue((Double) rawObject);
+    } else if (rawObject instanceof String) {
+      valueBuilder.setStringValue((String) rawObject);
+    } else if (rawObject instanceof Boolean) {
+      valueBuilder.setBoolValue((Boolean) rawObject);
+    } else if (rawObject instanceof Map) {
+      Struct.Builder structBuilder = Struct.newBuilder();
+      @SuppressWarnings("unchecked")
+      Map<String, ?> map = (Map<String, ?>) rawObject;
+      for (Map.Entry<String, ?> entry : map.entrySet()) {
+        structBuilder.putFields(entry.getKey(), convertToValue(entry.getValue()));
+      }
+      valueBuilder.setStructValue(structBuilder);
+    } else if (rawObject instanceof List) {
+      ListValue.Builder listBuilder = ListValue.newBuilder();
+      List<?> list = (List<?>) rawObject;
+      for (Object obj : list) {
+        listBuilder.addValues(convertToValue(obj));
+      }
+      valueBuilder.setListValue(listBuilder);
+    }
+    return valueBuilder.build();
+  }
+
+  enum ResourceType {
+    UNKNOWN, LDS, RDS, CDS, EDS;
+
+    String typeUrl() {
+      switch (this) {
+        case LDS:
+          return ADS_TYPE_URL_LDS;
+        case RDS:
+          return ADS_TYPE_URL_RDS;
+        case CDS:
+          return ADS_TYPE_URL_CDS;
+        case EDS:
+          return ADS_TYPE_URL_EDS;
+        case UNKNOWN:
+        default:
+          throw new AssertionError("Unknown or missing case in enum switch: " + this);
+      }
+    }
+
+    String typeUrlV2() {
+      switch (this) {
+        case LDS:
+          return ADS_TYPE_URL_LDS_V2;
+        case RDS:
+          return ADS_TYPE_URL_RDS_V2;
+        case CDS:
+          return ADS_TYPE_URL_CDS_V2;
+        case EDS:
+          return ADS_TYPE_URL_EDS_V2;
+        case UNKNOWN:
+        default:
+          throw new AssertionError("Unknown or missing case in enum switch: " + this);
+      }
+    }
+
+    static ResourceType fromTypeUrl(String typeUrl) {
+      switch (typeUrl) {
+        case ADS_TYPE_URL_LDS:
+          // fall trough
+        case ADS_TYPE_URL_LDS_V2:
+          return LDS;
+        case ADS_TYPE_URL_RDS:
+          // fall through
+        case ADS_TYPE_URL_RDS_V2:
+          return RDS;
+        case ADS_TYPE_URL_CDS:
+          // fall through
+        case ADS_TYPE_URL_CDS_V2:
+          return CDS;
+        case ADS_TYPE_URL_EDS:
+          // fall through
+        case ADS_TYPE_URL_EDS_V2:
+          return EDS;
+        default:
+          return UNKNOWN;
+      }
+    }
+  }
+
+  /**
+   * See corresponding Envoy proto message {@link
+   * io.envoyproxy.envoy.service.discovery.v3.DiscoveryResponse}.
+   */
+  static final class DiscoveryResponse {
```

Comment:
Ah... We should not do things like this. The xDS requests/responses have no reason to be defined/converted separately outside XdsClientImpl, all the processing for them are completely within the scope of XdsClientImpl. The converted data has no usage for gRPC, makes no sense to be here. How XdsClientImpl processes requests/responses should be encapsulated.

I see a bunch of `Any`, `MessageOrBuilder` inside, this newly defined datatype makes no sense that takes in a blob and itself is still a blob.